File: dotnet/mission-to-mars/Sonde.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace mission_to_mars
{
    public class Sonde
    {
        public void PreparerRécupération(IModuleMartien module)
        {
            if(module is Helicoptere)
            {
                PoserHelicoptere(module);
            }

            module.ActiverRecuperation();
        }

        private void PoserHelicoptere(IModuleMartien module)
        {
            int altitude = module.Position.Z;
            Atterir(module, altitude);
        }

        private void Atterir(IModuleMartien module, int altitude)
        {
            foreach (var _ in Enumerable.Range(0, altitude))
            {
                module.Descendre();
            }
        }
    }
}

File: python/mission-to-mars/src/mars/Sonde.py
from src.mars.Helicoptere import Helicoptere
from src.mars.ModuleMartien import ModuleMartien


class Sonde:

    def preparer_recuperation(self, module_martien: ModuleMartien) -> None:
        if isinstance(module_martien, Helicoptere):
            self._poser_helicoptere(module_martien)
        module_martien.activer_recuperation()

    def _poser_helicoptere(self, module_martien: ModuleMartien) -> None:
        altitude = module_martien.get_position().z
        self._atterir(module_martien, altitude)

    def _atterir(self, module_martien: ModuleMartien, altitude: int) -> None:
        for i in range(altitude):
            module_martien.descendre()
